File: examples/web_socket_commands.rb
require 'rubygems'
require 'em-websocket'

#this example must be used with the web_socket_command.js file.

EventMachine.run {
  @channel = EM::Channel.new

  EventMachine::WebSocket.start(:host => "0.0.0.0", :port => 8080, :debug => true) do |ws|

    ws.onopen {
      sid = @channel.subscribe { |msg| ws.send msg }
      @channel.push "#{sid} connected!"

      ws.onmessage { |msg|
        
        @channel.push ws.js_command('alert', 'test')
        #calls alert('test');
        
        @channel.push ws.js_command('prompt', 'prompt test', 9)
        #calls prompt('prompt test', 9);
        
        #javascript is a special command where the rest of the arguments are just evaluated.
        @channel.push ws.js_command('javascript', "alert('javascript alert')", "testval = 9+1", "alert(testval)")
        #runs the script below on the browser
        #     alert('javascript alert');
        #     testval = 9+1;
        #     alert(testval);
      }

      ws.onclose {
        @channel.unsubscribe(sid)
      }

    }
  end

  puts "Server started"
}

File: examples/js/web_socket_commands.js

var allowed_commands = ['alert', 'javascript', 'prompt'];

$(document).ready(function(){
  function debug(str){ $("#debug").append("<p>"+str+"</p>"); };

  ws = new WebSocket("ws://localhost:8080");
  ws.onmessage = function(evt) { 
    socket_command(evt.data);
    $("#msg").append("<p>"+evt.data+"</p>"); 
  };
  ws.onclose = function() { debug("socket closed"); };
  ws.onopen = function() {
    debug("connected...");
    ws.send("hello server");
  };
});

function socket_command(msg){
  try{
    jsonmsg = JSON.parse(msg);
    vars = jsonmsg.vars;
    command = jsonmsg.method_name;
  }catch(err){
    command = "";
    vars = [];
  }
  if(allowed_commands.contains(command)){
    //javascript is a special command where the rest of the arguments are just evaluated.
    if(command=='javascript'){
      for(x in vars){
        eval(vars[x].value);
      }
    }else{
      vars_array = [];
      for(x in vars){
        //right now we only have String and not NotString
        switch(vars[x].type){
          case 'String':
            vars_array.push("'"+vars[x].value+"'");
            break;
          default:
            vars_array.push(""+vars[x].value+"");
        }
      }
      eval(command+"("+vars_array.join(",")+")");
    }
  }else{
    $("#debug").append("<p>"+msg+"</p>");
  }
}
Array.prototype.contains = function(obj) {
  var i = this.length;
  while (i--) {
    if (this[i] === obj) {
      return true;
    }
  }
  return false;
}